[C10D] Document object collectives limitations

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #150880
* __->__ #150815

Adds louder warning labels in the doc page and docstring for object
collectives in hopes of raising awareness of several footgun issues
including accidental creation of cuda contexts by serializing and
sending 'device-local' gpu tensors over the object-* apis.


Preview:
<img width="902" alt="image" src="https://github.com/user-attachments/assets/e0c08c70-d8e5-4e15-b3e2-5cd563714f71" />


addresses #150798

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @d4l3k

wconstab committed:
[C10D] Document object collectives limitations

Adds louder warning labels in the doc page and docstring for object
collectives in hopes of raising awareness of several footgun issues
including accidental creation of cuda contexts by serializing and
sending 'device-local' gpu tensors over the object-* apis.

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Document object collectives limitations"


Adds louder warning labels in the doc page and docstring for object
collectives in hopes of raising awareness of several footgun issues
including accidental creation of cuda contexts by serializing and
sending 'device-local' gpu tensors over the object-* apis.


Preview:
<img width="902" alt="image" src="https://github.com/user-attachments/assets/e0c08c70-d8e5-4e15-b3e2-5cd563714f71" />


addresses #150798

cc H-Huang awgu wanchaol fegin fduwjj wz337 d4l3k

[ghstack-poisoned]

wconstab commented:
@pytorchbot help

wconstab committed:
Update on "[C10D] Document object collectives limitations"


Adds louder warning labels in the doc page and docstring for object
collectives in hopes of raising awareness of several footgun issues
including accidental creation of cuda contexts by serializing and
sending 'device-local' gpu tensors over the object-* apis.


Preview:
<img width="902" alt="image" src="https://github.com/user-attachments/assets/e0c08c70-d8e5-4e15-b3e2-5cd563714f71" />


addresses #150798

cc H-Huang awgu wanchaol fegin fduwjj wz337 d4l3k

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Document object collectives limitations"


Adds louder warning labels in the doc page and docstring for object
collectives in hopes of raising awareness of several footgun issues
including accidental creation of cuda contexts by serializing and
sending 'device-local' gpu tensors over the object-* apis.


Preview:
<img width="902" alt="image" src="https://github.com/user-attachments/assets/e0c08c70-d8e5-4e15-b3e2-5cd563714f71" />


addresses #150798

cc H-Huang awgu wanchaol fegin fduwjj wz337 d4l3k

[ghstack-poisoned]

kwen2501 commented on docs/source/distributed.rst:
```diff
@@ -554,7 +554,47 @@ Multi-GPU collective functions
+4. communicate the object data (second collective operation)
+5. convert raw data back into python (unpickle)
+
+Object collectives often have surprising performance or memory characteristics that lead to
```
"often" may be too heavy? :)

wconstab replied:
changed to "sometimes"

kwen2501 commented on docs/source/distributed.rst:
```diff
@@ -554,7 +554,47 @@ Multi-GPU collective functions
+
+**Asymetric pickle/unpickle time** - Pickling objects can be slow, depending on the number, type and size of the objects.
+When the collective has a fan-in (e.g. gather_object), the recieving rank(s) must unpickle N times more objects than
+the sending rank(s) had to pickle.  This has led to one class of issues where users report nccl watchdog timeouts caused
+by one rank spending a really long time unpickling data.
```
Why would the unpickle time lead to watchdog timeout? I thought the collective has finished at that point? Or are you saying that it may impact the next collective because this gathering rank would join late?

wconstab replied:
yea, its probably always an unrelated collective that actually times out, caused by desync, cuased by unpickling

kwen2501 commented on docs/source/distributed.rst:
```diff
@@ -554,7 +554,47 @@ Multi-GPU collective functions
+the sending rank(s) had to pickle.  This has led to one class of issues where users report nccl watchdog timeouts caused
+by one rank spending a really long time unpickling data.
+
+**Inefficient tensor communication** - Tensors should be sent via regular collective APIs, not object collective APIs.
+It is possible to send Tensors via object collective APIs, but they will be serialized and deserialized (including a
+cpu-sync and device-to-host copy in the case of non-cpu tensors), and in almost every case other than debugging or
+troubleshooting code, it would be worth the trouble to refactor the code to use non-object collectives instead.
+
+**Unexpected tensor devices** - If you still want to send tensors via object collectives, there is another aspect
+specific to cuda (and possibly other accelerators) tensors.  If you pickle a tensor that is currently on `cuda:3`, and
+then unpickle it, you will get another tensor on `cuda:3` *regardless of which process you are on, or which cuda device
+is the 'default' device for that process*.  With regular tensor collective APIs, 'output tensors' will always be on the
+same, local device, which is generally what you'd expect.
```
nit: maybe merge "Inefficient tensor communication" and "Unexpected tensor devices" into one section and name it "Not for tensors"

wconstab replied:
i merged the latter two instead. i think inefficient tensor comm applies to cpu and cuda tensors (any tensors) due to pickling overhead, and the latter sections apply to cuda specifically.

kwen2501 commented on docs/source/distributed.rst:
```diff
@@ -554,7 +554,47 @@ Multi-GPU collective functions
+
+**Unintended cuda-context creation** - Unpickling a tensor will implicity activate a cuda context if it is the first
+time a GPU is used by the process, which can waste significant amounts of GPU memory. This issue can be avoided by
+moving tensors to CPU before passing them as inputs to an object collective.
```
It seems "Unintended cuda-context creation" is really related to "Unexpected tensor devices" above. So I recommend merging sections too. 
Could the sentence "This issue can be avoided by moving tensors to CPU before passing them as inputs to an object collective" sound like we encourage the users to do so (while we actually don't)?

wconstab replied:
well, i said "worked around" not "should".  I think that the best I can do is outline all the risks, and if you still want to pass tensors in we can at least make it less bad.

kwen2501 reviewed (approved):
Thanks much for writing the paragraph! We need more foundational work like this!

awgu commented on docs/source/distributed.rst:
```diff
@@ -554,7 +554,47 @@ Multi-GPU collective functions
+Object collectives often have surprising performance or memory characteristics that lead to
+long runtimes or OOMs, and thus they should be used with caution.  Here are some common issues.
+
+**Asymetric pickle/unpickle time** - Pickling objects can be slow, depending on the number, type and size of the objects.
```
```suggestion
**Asymmetric pickle/unpickle time** - Pickling objects can be slow, depending on the number, type and size of the objects.
```

awgu commented on docs/source/distributed.rst:
```diff
@@ -554,7 +554,47 @@ Multi-GPU collective functions
+long runtimes or OOMs, and thus they should be used with caution.  Here are some common issues.
+
+**Asymetric pickle/unpickle time** - Pickling objects can be slow, depending on the number, type and size of the objects.
+When the collective has a fan-in (e.g. gather_object), the recieving rank(s) must unpickle N times more objects than
```
```suggestion
When the collective has a fan-in (e.g. gather_object), the receiving rank(s) must unpickle N times more objects than
```

awgu commented on docs/source/distributed.rst:
```diff
@@ -554,7 +554,47 @@ Multi-GPU collective functions
+
+**Asymetric pickle/unpickle time** - Pickling objects can be slow, depending on the number, type and size of the objects.
+When the collective has a fan-in (e.g. gather_object), the recieving rank(s) must unpickle N times more objects than
+the sending rank(s) had to pickle.  This has led to one class of issues where users report nccl watchdog timeouts caused
```
```suggestion
the sending rank(s) had to pickle.  This has led to one class of issues where users report NCCL watchdog timeouts caused
```

awgu commented on docs/source/distributed.rst:
```diff
@@ -554,7 +554,47 @@ Multi-GPU collective functions
+is the 'default' device for that process*.  With regular tensor collective APIs, 'output tensors' will always be on the
+same, local device, which is generally what you'd expect.
+
+**Unintended cuda-context creation** - Unpickling a tensor will implicity activate a cuda context if it is the first
```
```suggestion
**Unintended cuda-context creation** - Unpickling a tensor will implicitly activate a CUDA context if it is the first
```

awgu commented on docs/source/distributed.rst:
```diff
@@ -554,7 +554,47 @@ Multi-GPU collective functions
+
+**Inefficient tensor communication** - Tensors should be sent via regular collective APIs, not object collective APIs.
+It is possible to send Tensors via object collective APIs, but they will be serialized and deserialized (including a
+cpu-sync and device-to-host copy in the case of non-cpu tensors), and in almost every case other than debugging or
```
```suggestion
CPU-sync and device-to-host copy in the case of non-CPU tensors), and in almost every case other than debugging or
```

awgu commented on docs/source/distributed.rst:
```diff
@@ -554,7 +554,47 @@ Multi-GPU collective functions
+    if they are safe to use for your use case.
+
+Object collectives are a set of collective-like operations that work on arbitrary
+python objects, as long as they can be pickled.  There are various collective patterns
```
```suggestion
Python objects, as long as they can be pickled.  There are various collective patterns
```

awgu commented on docs/source/distributed.rst:
```diff
@@ -554,7 +554,47 @@ Multi-GPU collective functions
+2. communicate the size of this byte tensor to peers (first collective operation)
+3. allocate appropriately sized tensor to perform the real collective
+4. communicate the object data (second collective operation)
+5. convert raw data back into python (unpickle)
```
```suggestion
5. convert raw data back into Python (unpickle)
```

awgu commented on docs/source/distributed.rst:
```diff
@@ -554,7 +554,47 @@ Multi-GPU collective functions
+
+**Unexpected tensor devices** - If you still want to send tensors via object collectives, there is another aspect
+specific to cuda (and possibly other accelerators) tensors.  If you pickle a tensor that is currently on `cuda:3`, and
+then unpickle it, you will get another tensor on `cuda:3` *regardless of which process you are on, or which cuda device
```
```suggestion
then unpickle it, you will get another tensor on `cuda:3` *regardless of which process you are on, or which CUDA device
```

awgu reviewed (commented):
proof-reading-only pass

(some typos and python/cpu/cuda -> Python/CPU/CUDA)

wconstab committed:
Update on "[C10D] Document object collectives limitations"


Adds louder warning labels in the doc page and docstring for object
collectives in hopes of raising awareness of several footgun issues
including accidental creation of cuda contexts by serializing and
sending 'device-local' gpu tensors over the object-* apis.


Preview:
<img width="902" alt="image" src="https://github.com/user-attachments/assets/e0c08c70-d8e5-4e15-b3e2-5cd563714f71" />


addresses #150798

cc H-Huang awgu wanchaol fegin fduwjj wz337 d4l3k

[ghstack-poisoned]

svekars commented on torch/distributed/distributed_c10d.py:
```diff
@@ -3078,6 +3078,10 @@ def all_gather_object(object_list, obj, group=None):
 
+    .. warning::
+        Object collectives have a number of serious performance and scalability
+        limitations.  See :ref:`_object_collectives` for details.
```
```suggestion
        limitations.  See :ref:`object_collectives` for details.
```

wconstab replied:
Thanks @svekars - this solved my issue

wconstab committed:
Update on "[C10D] Document object collectives limitations"


Adds louder warning labels in the doc page and docstring for object
collectives in hopes of raising awareness of several footgun issues
including accidental creation of cuda contexts by serializing and
sending 'device-local' gpu tensors over the object-* apis.


Preview:
<img width="902" alt="image" src="https://github.com/user-attachments/assets/e0c08c70-d8e5-4e15-b3e2-5cd563714f71" />


addresses #150798

cc H-Huang awgu wanchaol fegin fduwjj wz337 d4l3k

[ghstack-poisoned]

wconstab commented:
@pytorchbot merge
